clients/da-vinci-client/src/main/java/com/linkedin/davinci/blob/BlobSnapshotManager.java

```diff
@@ -0,0 +1,139 @@
+   * otherwise it increases the count of people using the snapshot
+   */
+  public void maybeUpdateHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    snapShotTimestamps.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
+    concurrentSnapshotUsers.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
+    concurrentSnapshotUsers.get(topicName).putIfAbsent(partitionId, new AtomicLong(0));
+    try (AutoCloseableLock ignored = AutoCloseableLock.of(lock)) {
+      if (isSnapshotStale(topicName, partitionId)) {
+        if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
+            || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
+          updateHybridSnapshot(rocksDB, topicName, partitionId);
+          snapShotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
+        }
+      }
+      concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
+      LOGGER.info(
+          "Retrieved snapshot from {} with timestamp {}",
+          fullPathForPartitionDBSnapshot,
+          snapShotTimestamps.get(topicName).get(partitionId));
+    }
+  }
+
+  public void decreaseConcurrentUserCount(String topicName, int partitionId) {
+    Map<Integer, AtomicLong> concurrentPartitionUsers = concurrentSnapshotUsers.get(topicName);
+    if (concurrentPartitionUsers == null) {
+      throw new VeniceException("No topic found: " + topicName);
+    }
+    AtomicLong concurrentUsers = concurrentPartitionUsers.get(partitionId);
+    if (concurrentUsers == null) {
+      throw new VeniceException(String.format("%d partition not found on topic %s", partitionId, topicName));
+    }
+    try (AutoCloseableLock ignored = AutoCloseableLock.of(lock)) {
+      long result = concurrentUsers.decrementAndGet();
+      if (result < 0) {
+        throw new VeniceException("Concurrent user count cannot be negative");
+      }
+    }
+  }
+
+  long getConcurrentSnapshotUsers(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+    if (!concurrentSnapshotUsers.containsKey(topicName)
+        || !concurrentSnapshotUsers.get(topicName).containsKey(partitionId)) {
+      return 0;
+    }
+    return concurrentSnapshotUsers.get(topicName).get(partitionId).get();
+  }
+
+  /**
+   * Checks if the current snapshot of the partition is stale
+   */
+  private boolean isSnapshotStale(String topicName, int partitionId) {
+    if (!snapShotTimestamps.containsKey(topicName) || !snapShotTimestamps.get(topicName).containsKey(partitionId)) {
```

Comment:
I think we missed one more thing - check if the store is hybrid or not. If the store is batch only, we should not do useless checkpoint update.

What we can do is: Take in a ReadOnlyStoreRepository in the constructor, and in this method, check if store's hybrid config is null or not, if it is null, then it is a batch store, then you should just return false.

Reply:
Also, could you add a simple unit test about this? Say if you test against a batch store, then the updateHybridSnapshot invocation count should be 0, even you intentionally set the snapshot TS to be stale.